File: apps/ios/TePlannerApp/CapabilityParamEditor.swift
import SwiftUI
import TePlannerKit

/// Per-capability inline form rows shown in the rule builder when
/// the user has picked a capability for a `notify_and_offer` action.
/// Extracted from RuleBuilderView so each capability's UI lives in
/// one place and can grow without bloating the builder file.
///
/// The view binds to the rule's `paramOverrides` dict (as
/// Binding<[String: JSONValue]>) — every row mutates a single key
/// inside that dict. When a capability id has no specific UI here
/// we render `无需参数` so the user knows nothing is missing.
struct CapabilityParamEditor: View {
    let capabilityId: String
    @Binding var paramOverrides: [String: JSONValue]

    var body: some View {
        switch capabilityId {
        case "tesla.climate.set_keeper_mode":
            Picker("模式", selection: intBinding("mode", default: 0)) {
                Text("关闭").tag(0); Text("保持").tag(1)
                Text("宠物模式").tag(2); Text("露营模式").tag(3)
            }
        case "tesla.security.set_sentry":
            Toggle("开启哨兵", isOn: boolBinding("on", default: false))
        case "tesla.charging.set_limit":
            Stepper(value: intBinding("percent", default: 80), in: 50...100, step: 5) {
                HStack { Text("限额"); Spacer()
                    Text("\(intBinding("percent", default: 80).wrappedValue)%")
                        .foregroundStyle(.secondary).monospacedDigit()
                }
            }
        case "tesla.charging.set_amps":
            Stepper(value: intBinding("amps", default: 16), in: 5...48) {
                HStack { Text("电流"); Spacer()
                    Text("\(intBinding("amps", default: 16).wrappedValue) A")
                        .foregroundStyle(.secondary).monospacedDigit()
                }
            }
        case "tesla.climate.set_temps":
            HStack { Text("主驾"); Spacer()
                Stepper(value: doubleBinding("driver_temp", default: 22), in: 15...28, step: 0.5) {
                    Text("\(doubleBinding("driver_temp", default: 22).wrappedValue, specifier: "%.1f") °C")
                        .foregroundStyle(.secondary)
                }
            }
            HStack { Text("副驾"); Spacer()
                Stepper(value: doubleBinding("passenger_temp", default: 22), in: 15...28, step: 0.5) {
                    Text("\(doubleBinding("passenger_temp", default: 22).wrappedValue, specifier: "%.1f") °C")
                        .foregroundStyle(.secondary)
                }
            }
        case "tesla.climate.set_preconditioning_max":
            Toggle("开启最大预热", isOn: boolBinding("on", default: true))
        case "tesla.climate.set_cabin_overheat":
            Picker("模式", selection: intBinding("mode", default: 2)) {
                Text("关闭").tag(0); Text("空调").tag(1); Text("仅风扇").tag(2)
            }
        case "tesla.comfort.set_seat_heater":
            Picker("座位", selection: intBinding("seat", default: 0)) {
                Text("主驾").tag(0); Text("副驾").tag(1)
                Text("后排左").tag(2); Text("后排中").tag(4); Text("后排右").tag(5)
            }
            Picker("档位", selection: intBinding("level", default: 2)) {
                Text("关闭").tag(0); Text("低").tag(1); Text("中").tag(2); Text("高").tag(3)
            }
        case "tesla.comfort.set_steering_wheel_heater":
            Toggle("开启方向盘加热", isOn: boolBinding("on", default: true))
        case "tesla.media.set_volume":
            HStack { Text("音量"); Spacer()
                Stepper(value: doubleBinding("volume", default: 5), in: 0...11, step: 0.5) {
                    Text("\(doubleBinding("volume", default: 5).wrappedValue, specifier: "%.1f")")
                        .foregroundStyle(.secondary)
                }
            }
        default:
            Text("无需参数")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
    }

    private func intBinding(_ key: String, default def: Int) -> Binding<Int> {
        Binding(
            get: { paramOverrides[key]?.intValue ?? def },
            set: { paramOverrides[key] = .int($0) }
        )
    }

    private func boolBinding(_ key: String, default def: Bool) -> Binding<Bool> {
        Binding(
            get: { paramOverrides[key]?.boolValue ?? def },
            set: { paramOverrides[key] = .bool($0) }
        )
    }

    private func doubleBinding(_ key: String, default def: Double) -> Binding<Double> {
        Binding(
            get: { paramOverrides[key]?.doubleValue ?? def },
            set: { paramOverrides[key] = .double($0) }
        )
    }
}

File: apps/android/app/src/main/java/cloud/teplanner/android/automations/CapabilityParamEditor.kt
package cloud.teplanner.android.automations

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.FilledTonalIconButton
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull

/**
 * Mirror of iOS CapabilityParamEditor. For a given capability id,
 * render the right inline form rows so the user can override the
 * defaults from [CapabilityDefaults.params].
 *
 * Capabilities not enumerated here render "无需参数" — same as iOS
 * `default:` branch.
 */
@Composable
fun CapabilityParamEditor(
    capabilityId: String,
    params: JsonObject,
    onChange: (JsonObject) -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        when (capabilityId) {
            "tesla.climate.set_keeper_mode" ->
                IntPicker(
                    label = "模式", key = "mode", default = 0,
                    options = listOf(0 to "关闭", 1 to "保持", 2 to "宠物模式", 3 to "露营模式"),
                    params = params, onChange = onChange,
                )
            "tesla.security.set_sentry" ->
                BoolSwitch("开启哨兵", "on", false, params, onChange)
            "tesla.charging.set_limit" ->
                IntStepper(
                    label = "限额", key = "percent", default = 80,
                    min = 50, max = 100, step = 5, suffix = "%",
                    params = params, onChange = onChange,
                )
            "tesla.charging.set_amps" ->
                IntStepper(
                    label = "电流", key = "amps", default = 16,
                    min = 5, max = 48, step = 1, suffix = " A",
                    params = params, onChange = onChange,
                )
            "tesla.climate.set_temps" -> {
                DoubleStepper(
                    label = "主驾", key = "driver_temp", default = 22.0,
                    min = 15.0, max = 28.0, step = 0.5, suffix = " °C",
                    params = params, onChange = onChange,
                )
                DoubleStepper(
                    label = "副驾", key = "passenger_temp", default = 22.0,
                    min = 15.0, max = 28.0, step = 0.5, suffix = " °C",
                    params = params, onChange = onChange,
                )
            }
            "tesla.climate.set_preconditioning_max" ->
                BoolSwitch("开启最大预热", "on", true, params, onChange)
            "tesla.climate.set_cabin_overheat" ->
                IntPicker(
                    label = "模式", key = "mode", default = 2,
                    options = listOf(0 to "关闭", 1 to "空调", 2 to "仅风扇"),
                    params = params, onChange = onChange,
                )
            "tesla.comfort.set_seat_heater" -> {
                IntPicker(
                    label = "座位", key = "seat", default = 0,
                    options = listOf(
                        0 to "主驾", 1 to "副驾",
                        2 to "后排左", 4 to "后排中", 5 to "后排右",
                    ),
                    params = params, onChange = onChange,
                )
                IntPicker(
                    label = "档位", key = "level", default = 2,
                    options = listOf(0 to "关闭", 1 to "低", 2 to "中", 3 to "高"),
                    params = params, onChange = onChange,
                )
            }
            "tesla.comfort.set_steering_wheel_heater" ->
                BoolSwitch("开启方向盘加热", "on", true, params, onChange)
            "tesla.media.set_volume" ->
                DoubleStepper(
                    label = "音量", key = "volume", default = 5.0,
                    min = 0.0, max = 11.0, step = 0.5, suffix = "",
                    params = params, onChange = onChange,
                )
            else -> Text(
                "无需参数",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}


@Composable
private fun IntPicker(
    label: String,
    key: String,
    default: Int,
    options: List<Pair<Int, String>>,
    params: JsonObject,
    onChange: (JsonObject) -> Unit,
) {
    val current = (params[key] as? JsonPrimitive)?.intOrNull ?: default
    Column {
        Text(label, style = MaterialTheme.typography.bodyMedium)
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            options.forEach { (v, lbl) ->
                FilterChip(
                    selected = current == v,
                    onClick = { onChange(setInt(params, key, v)) },
                    label = { Text(lbl) },
                )
            }
        }
    }
}


@Composable
private fun BoolSwitch(
    label: String,
    key: String,
    default: Boolean,
    params: JsonObject,
    onChange: (JsonObject) -> Unit,
) {
    val current = (params[key] as? JsonPrimitive)?.booleanOrNull ?: default
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, modifier = Modifier.weight(1f))
        Switch(
            checked = current,
            onCheckedChange = { onChange(setBool(params, key, it)) },
        )
    }
}


@Composable
private fun IntStepper(
    label: String, key: String, default: Int,
    min: Int, max: Int, step: Int, suffix: String,
    params: JsonObject, onChange: (JsonObject) -> Unit,
) {
    val value = (params[key] as? JsonPrimitive)?.intOrNull ?: default
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, modifier = Modifier.weight(1f))
        FilledTonalIconButton(
            onClick = { if (value > min) onChange(setInt(params, key, value - step)) },
            modifier = Modifier.size(32.dp),
            enabled = value > min,
        ) { Icon(Icons.Filled.Remove, "减") }
        Box(Modifier.width(72.dp), contentAlignment = Alignment.Center) {
            Text(
                "$value$suffix",
                style = MaterialTheme.typography.titleSmall,
            )
        }
        FilledTonalIconButton(
            onClick = { if (value < max) onChange(setInt(params, key, value + step)) },
            modifier = Modifier.size(32.dp),
            enabled = value < max,
        ) { Icon(Icons.Filled.Add, "加") }
    }
}


@Composable
private fun DoubleStepper(
    label: String, key: String, default: Double,
    min: Double, max: Double, step: Double, suffix: String,
    params: JsonObject, onChange: (JsonObject) -> Unit,
) {
    val value = (params[key] as? JsonPrimitive)?.doubleOrNull ?: default
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, modifier = Modifier.weight(1f))
        FilledTonalIconButton(
            onClick = { if (value > min) onChange(setDouble(params, key, value - step)) },
            modifier = Modifier.size(32.dp),
            enabled = value > min,
        ) { Icon(Icons.Filled.Remove, "减") }
        Box(Modifier.width(72.dp), contentAlignment = Alignment.Center) {
            Text(
                "%.1f%s".format(value, suffix),
                style = MaterialTheme.typography.titleSmall,
            )
        }
        FilledTonalIconButton(
            onClick = { if (value < max) onChange(setDouble(params, key, value + step)) },
            modifier = Modifier.size(32.dp),
            enabled = value < max,
        ) { Icon(Icons.Filled.Add, "加") }
    }
}


private fun setInt(obj: JsonObject, key: String, v: Int): JsonObject =
    buildJsonObject {
        obj.forEach { (k, value) -> if (k != key) put(k, value) }
        put(key, JsonPrimitive(v))
    }

private fun setBool(obj: JsonObject, key: String, v: Boolean): JsonObject =
    buildJsonObject {
        obj.forEach { (k, value) -> if (k != key) put(k, value) }
        put(key, JsonPrimitive(v))
    }

private fun setDouble(obj: JsonObject, key: String, v: Double): JsonObject =
    buildJsonObject {
        obj.forEach { (k, value) -> if (k != key) put(k, value) }
        put(key, JsonPrimitive(v))
    }
